### ISSUE TITLE
`prevent-abbreviations` does not work with object destructuring

### ISSUE THREAD

**Issue by ehmicky** [sindresorhus/eslint-plugin-unicorn#270]
`a.js`:

```js
const exampleFunction = function({ opts }) {}
const { opts } = object
```

```bash
$ eslint a.js
/home/user/a.js
  1:42  error  The variable `opts` should be named `options`. A more descriptive name will do too  unicorn/prevent-abbreviations
  2:15  error  The variable `opts` should be named `options`. A more descriptive name will do too  unicorn/prevent-abbreviations

✖ 2 problems (2 errors, 0 warnings)
  2 errors and 0 warnings potentially fixable with the `--fix` option.
```

`b.js`:

```js
const exampleFunction = function({ opts: { example } }) {}
const { opts: { example } } = object
```

```bash
$ eslint b.js
```

`b.js` does not report any error but it should. 

# Configuration

`.eslintrc.json`:

```json
{
  "parserOptions": {
    "ecmaVersion": 2019
  },
  "plugins": ["unicorn"],
  "rules": {
    "unicorn/prevent-abbreviations": ["error", { "checkShorthandProperties": true }]
  }
}
```

# Environment

```bash
$ npm ls eslint eslint-plugin-unicorn
desktop@1.0.0 /home/user/Desktop
├── eslint@5.15.3 
└── eslint-plugin-unicorn@8.0.0

$ node --version
v11.12.0

$ lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 18.10
Release:	18.10
Codename:	cosmic
```

**Comment by futpib** [comment]
Have you tried [`checkShorthandProperties`](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/master/docs/rules/prevent-abbreviations.md#checkshorthandproperties) option?

**Comment by ehmicky** [comment]
@futpib Thanks! Now actually it turns out my problem is about object destructuring. I've updated my initial message.

**Comment by fisker** [comment]
I think this works as expected

```js
const { opts: { example } } = object
```

means

```js
const example = object.opts.example
```

the key `opts` can't change, this is not a bug

**Comment by ehmicky** [comment]
The `opts` can't be change indeed, i.e. it cannot be automatically fixed. However this could still be reported as an error.

**Comment by fisker** [comment]
have a look at #271 , I think It's the default options problem.

**Comment by ehmicky** [comment]
@fisker I don't think those issues are related.

**Comment by fisker** [comment]
I see what's your point now, you mean when destructuring a object, the abbr prop should report a error.

I can't agree with this, code doesn't know where your object come from

here is a example, ESLint rule expose a object 

```js
{
    meta: {
        type: "suggestion",
        docs: {
            description: "disallow unnecessary semicolons",
            category: "Possible Errors",
            recommended: true,
            url: "https://eslint.org/docs/rules/no-extra-semi"
        },
        fixable: "code",
        schema: [] // no options
    },
}
```
https://eslint.org/docs/developer-guide/working-with-rules

you can't force another program use `documents` instead of `docs`, 

to get meta.docs.url we can do something like,  If ther is a error, how will you fix it ? 

```js
const {meta:{docs: { url }}}= obj

// or
const {docs: { url }}= obj.meta

// or
const { url }= obj.meta.docs
```

this could be better, and this is want I'm using, but do you want this also need report a error ?

```js
const {docs: documents = {}} = meta
const {url} = documents
```

**Comment by ehmicky** [comment]
For variables that cannot be modified (e.g. because they come from a library), users can either:
  - add a `eslint-disable` comment
  - use the rule whitelisting options. This rule has several such options to cover this.

The problem you're pointing out is not specific to this ESLint rule, but is a more generic ESLint problem. For example, this could be said of several core ESLint rules related to variable naming such as [`id-length`](https://eslint.org/docs/rules/id-length) or [`id-blacklist`](https://eslint.org/docs/rules/id-blacklist):

```json
{
	"parserOptions": {
	  "ecmaVersion": 2019
  },
  "rules": {
    "id-blacklist": ["error", "test"],
    "id-length": ["error", { "max": 2 }]
  }
}
```

On the following file:

```js
const { test: t } = require('test')
```

Prints:

```bash
/home/user/a.js
  1:9  error  Identifier 'test' is blacklisted          id-blacklist
  1:9  error  Identifier name 'test' is too long (> 2)  id-length

✖ 2 problems (2 errors, 0 warnings)
```

**Comment by sindresorhus** [comment]
I would recommend reading through the discussions in https://github.com/sindresorhus/eslint-plugin-unicorn/pull/237. There were a lot of considerations to take into account. We tried to find a balance of being helpful but not annoying.

I generally don't mind making the rule stricter than now, but it should be opt-in through options.

Someone needs to make a proposal on exactly what additional options should be introduced.

**Comment by ehmicky** [comment]
Yes I see, it's quite complicated. I can close this issue if you think this is better.

**Comment by sindresorhus** [comment]
I don't really have time to make a proposal, but we can keep it open in case anyone wants to make a change proposal.

**Comment by fregante** [comment]
Closing for now, no movement in 3+ years
